### PR TITLE
Add LoRA as efficient finetuning strategy

### PR DESCRIPTION
**Pull request by Raldir**
Added the option to apply LoRA (https://arxiv.org/pdf/2106.09685.pdf) to Huggingface text model as an additional option for efficient fine-tuning, see PR #1666. New PR based on #1780 for easier merge. Code is largely based on https://github.com/microsoft/LoRA/blob/main/loralib/layers.py

- Added flag to optimizer `optimization.efficient_finetuned`: `lora`, `lora_bias` and `lora_norm`. The first fine-tunes only loRA weights, the second updates both LoRA weights and all biases of the transformer (LoRA + bit_fit), and the third updates LoRA, all biases and normalization terms (LoRA + norm_fit).
- LoRA uses three hyperparameters. These are also set through optimization config. While Lora `r` and `alpha` are quite robust and likely do not need much tuning, the `filter` parameter restricts which linear layers LoRA should be injected into. The original paper indicates that Query and Value attention weights work best so this is set as the default for now. Depending on use we might need to consider a more sophisticated filter algorithm, e.g. using regular expressions.   
- Currently, when saving the model the entire model is stored, which is unnecessary. TODO: Need to change to solely save fine-tuned weights, also applies to bit_fit and norm_fit.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: text/src/autogluon/text/automm/utils.py [hunk 1]
```diff
@@ -764,6 +764,9 @@ def create_model(
         else:
             raise ValueError(f"unknown model name: {model_name}")
 
+        if config.optimization.efficient_finetune:
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/utils.py:767]: Change to the following:

```python
if OmegaConf.select(config, 'optimization.efficient_finetune')
```

File: text/src/autogluon/text/automm/utils.py [hunk 2]
```diff
@@ -775,6 +778,29 @@ def create_model(
...
+    """
+    Apply an adaptation to the model for efficient fine-tuning.
+    Parameters
+    ----------
+    model
+        A PyTorch model.
+    config:
+        A DictConfig object. The optimization config should be accessible by "config.optimization".
+    """
+    if "lora" in config.optimization.efficient_finetune:
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/utils.py:791]: Change `config.optimization.efficient_finetune` to `OmegaConf.select(config, "optimization.efficient_finetune")`

**Comment by sxjscience** [comment]
Let me edit the file to see if the CI will pass.

**Comment by sxjscience** [comment]
Really appreciate the contribution!!! 

We can change the saving logic in a later PR. If the CI passes, the PR looks good to me. Some remaining items:

1) Profiling on text backbone other than mdeberta, e.g., roberta, electra, etc
2) Change the saving logic to solely save fine-tuned weights, also applies to bit_fit and norm_fit.

**Comment by szha** [comment]
Job PR-1809-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1809/6/index.html

File: text/src/autogluon/text/automm/utils.py [hunk 3]
```diff
@@ -775,6 +778,29 @@ def create_model(
         raise ValueError(f"No available models for {names}")
 
 
+def apply_model_adaptation(model: nn.Module, config: DictConfig) -> nn.Module:
+    """
+    Apply an adaptation to the model for efficient fine-tuning.
+    Parameters
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/utils.py:784]: A newline is required above the `Parameters`.

File: text/src/autogluon/text/automm/models/lora_layers.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                    self.sparse,
+                )
+                result += (after_A @ self.lora_B.T) * self.scaling
+            return result
+        else:
+            return nn.Embedding.forward(self, x)
+
+
+class MergedLinear(nn.Linear, LoRALayer):
+    # LoRA implemented in a dense layer
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/models/lora_layers.py:217]: Need docstrings.

File: text/src/autogluon/text/automm/models/lora_layers.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                after_A = F.linear(self.lora_dropout(x), self.lora_A)
+                after_B = F.conv1d(
+                    after_A.transpose(-2, -1), self.lora_B.unsqueeze(-1), groups=sum(self.enable_lora)
+                ).transpose(-2, -1)
+                result += self.zero_pad(after_B) * self.scaling
+            return result
+
+
+class Conv2d(nn.Conv2d, LoRALayer):
+    # LoRA implemented in a dense layer
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/models/lora_layers.py:311]: Need docstrings.

Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/models/lora_layers.py:311]: Is name `Conv2d` too general? Maybe consider make this name related to lora?

**Comment by szha** [comment]
Job PR-1809-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1809/7/index.html

**Review by zhiqiangdon — APPROVED** [review]
LGTM.
